package RemyBuffers;

message WhiskerTree {
  optional MemoryRange domain = 1;

  repeated WhiskerTree children = 2;

  optional Whisker leaf = 3;

  optional ConfigRange config = 4;

  optional OptimizationSettings optimizer = 5;

  optional ConfigVector configvector = 6;
}

message FinTree {
  optional MemoryRange domain = 90;

  repeated FinTree children = 91;

  optional Fin leaf = 92;

  optional ConfigRange config = 93;

  optional OptimizationSettings optimizer = 94;

  optional ConfigVector configvector = 95;
}

message MemoryRange {
  optional Memory lower = 11;
  optional Memory upper = 12;
  
  enum Axis {
    SEND_EWMA = 0;
    REC_EWMA = 1;
    RTT_RATIO = 2;
    SLOW_REC_EWMA = 3;
    RTT_DIFF = 4;
    QUEUEING_DELAY = 5;
  }

  repeated Axis active_axis = 13;
}

message Memory {
  optional double rec_send_ewma = 21;
  optional double rec_rec_ewma = 22;
  optional double rtt_ratio = 23;
  optional double slow_rec_rec_ewma = 24;
  optional double rtt_diff = 25;
  optional double queueing_delay = 26;
}

message Whisker {
  optional sint32 window_increment = 31;
  optional double window_multiple = 32;
  optional double intersend = 33;

  optional MemoryRange domain = 34;
}

message Fin {
  optional double lambda = 37;

  optional MemoryRange domain = 38;
}

message OptimizationSetting {
  optional double min_value = 41;
  optional double max_value = 42;

  optional double min_change = 43;
  optional double max_change = 44;

  optional double multiplier = 45;

  optional double default_value = 46;
}

message OptimizationSettings {
  optional OptimizationSetting window_increment = 51;
  optional OptimizationSetting window_multiple = 52;
  optional OptimizationSetting intersend = 53;
  optional OptimizationSetting lambda = 54;
}

message Range {
  optional double low = 61;
  optional double high = 62;
  optional double incr = 63;
}

message ConfigRange {
  optional Range link_packets_per_ms = 71;
  optional Range rtt = 72;
  optional Range num_senders = 73;
  optional Range buffer_size = 74;
  optional Range mean_off_duration = 75;
  optional Range mean_on_duration = 76;
  optional uint32 simulation_ticks = 77;
  optional Range stochastic_loss_rate = 78;
}

message NetConfig {
  optional double mean_on_duration = 1;
  optional double mean_off_duration = 2;
  optional uint32 num_senders = 3;
  optional double link_ppt = 4;
  optional double delay = 5;
  optional uint32 buffer_size = 6;
  optional double stochastic_loss_rate = 7;
}

message ConfigVector {
  repeated NetConfig config = 81;
}
